src/librustc/infer/lattice.rs

```diff
@@ -64,14 +65,14 @@ pub fn super_lattice_tys<'a, 'gcx, 'tcx, L>(this: &mut L,
     match (&a.sty, &b.sty) {
         (&ty::TyInfer(TyVar(..)), &ty::TyInfer(TyVar(..)))
             if infcx.type_var_diverges(a) && infcx.type_var_diverges(b) => {
-            let v = infcx.next_diverging_ty_var();
+            let v = infcx.next_diverging_ty_var(TypeVariableOrigin::LatticeVariable);
```

Comment:
Ack, the complication here is that `self.fields.cause` is an `Option<ObligationCause>`. How should I ensure that there must always be an `ObligationCause`?